File: clients_java/client/src/org/openamq/client/message/MessageFactoryRegistry.java
package org.openamq.client.message;

import org.openamq.AMQException;
import org.openamq.framing.JmsContentHeaderProperties;
import org.openamq.framing.ContentHeaderBody;

import javax.jms.JMSException;
import java.util.HashMap;
import java.util.Map;
import java.util.List;

/**
 * @author Robert Greig (robert.j.greig@jpmorgan.com)
 */
public class MessageFactoryRegistry
{
    private final Map _mimeToFactoryMap = new HashMap();

    public void registerFactory(String mimeType, MessageFactory mf)
    {
//        if (!FieldTable.grm && mimeType == null)
//        {
//            throw new IllegalArgumentException("Mime time must not be null");
//        }
        if (mf == null)
        {
            throw new IllegalArgumentException("Message factory must not be null");
        }
        _mimeToFactoryMap.put(mimeType, mf);
    }

    public MessageFactory deregisterFactory(String mimeType)
    {
        return (MessageFactory) _mimeToFactoryMap.remove(mimeType);
    }

    /**
     * Create a message. This looks up the MIME type from the content header and instantiates the appropriate
     * concrete message type.
     * @param messageNbr the AMQ message id
     * @param redelivered true if redelivered
     * @param contentHeader the content header that was received
     * @param bodies a list of ContentBody instances
     * @return the message.
     * @throws AMQException
     * @throws JMSException
     */
    public AbstractJMSMessage createMessage(long messageNbr, boolean redelivered,
                                            ContentHeaderBody contentHeader,
                                            List bodies) throws AMQException, JMSException
    {
        JmsContentHeaderProperties properties =  (JmsContentHeaderProperties) contentHeader.properties;
        MessageFactory mf = (MessageFactory) _mimeToFactoryMap.get(properties.contentType);
        if (mf == null)
        {
            throw new AMQException("Unsupport MIME type of " + properties.contentType);
        }
        else
        {
            return mf.createMessage(messageNbr, redelivered, contentHeader, bodies);
        }
    }

    public AbstractJMSMessage createMessage(String mimeType) throws AMQException, JMSException
    {
        if (mimeType == null)
        {
            throw new IllegalArgumentException("Mime type must not be null");
        }
        MessageFactory mf = (MessageFactory) _mimeToFactoryMap.get(mimeType);
        if (mf == null)
        {
            throw new AMQException("Unsupport MIME type of " + mimeType);
        }
        else
        {
            return mf.createMessage();
        }
    }

    /**
     * Construct a new registry with the default message factories registered
     * @return a message factory registry
     */
    public static MessageFactoryRegistry newDefaultRegistry()
    {
        MessageFactoryRegistry mf = new MessageFactoryRegistry();
        mf.registerFactory("text/plain", new JMSTextMessageFactory());
        mf.registerFactory("application/octet-stream", new JMSBytesMessageFactory());
        // TODO: use bytes message for default message factory
        //if (FieldTable.grm)
            mf.registerFactory(null, new JMSTextMessageFactory());
        return mf;
    }
}

File: dotnet/xmsclient/Client/Message/MessageFactoryRegistry.cs
using System;
using System.Collections;
using jpmorgan.mina.common;
using log4net;
using IBM.XMS;
using OpenAMQ.Framing;

namespace OpenAMQ.XMS.Client.Message
{
    public class MessageFactoryRegistry
    {
        private readonly Hashtable _mimeToFactoryMap = new Hashtable();

        public void RegisterFactory(string mimeType, IMessageFactory mf)
        {    
            if (mf == null)
            {
                throw new ArgumentNullException("Message factory");
            }
            if (mimeType == null)
            {
                throw new ArgumentNullException("mf");
            }
            _mimeToFactoryMap[mimeType] = mf;
        }

        public void DeregisterFactory(string mimeType)
        {
            _mimeToFactoryMap.Remove(mimeType);
        }

        /// <summary>
        /// Create a message. This looks up the MIME type from the content header and instantiates the appropriate
        /// concrete message type.
        /// </summary>
        /// <param name="messageNbr">the AMQ message id</param>
        /// <param name="redelivered">true if redelivered</param>
        /// <param name="contentHeader">the content header that was received</param>
        /// <param name="bodies">a list of ContentBody instances</param>
        /// <returns>the message.</returns>
        /// <exception cref="AMQException"/>
        /// <exception cref="XMSException"/>
        public AbstractXMSMessage CreateMessage(ulong messageNbr, bool redelivered,
                                                ContentHeaderBody contentHeader,
                                                IList bodies)
        {
            BasicContentHeaderProperties properties =  (BasicContentHeaderProperties) contentHeader.Properties;
            IMessageFactory mf = (IMessageFactory) _mimeToFactoryMap[properties.ContentType];
            if (mf == null)
            {
                throw new AMQException("Unsupport MIME type of " + properties.ContentType);
            }
            else
            {
                return mf.CreateMessage(messageNbr, redelivered, contentHeader, bodies);
            }
        }

        public AbstractXMSMessage CreateMessage(string mimeType)
        {
            if (mimeType == null)
            {
                throw new ArgumentNullException("Mime type must not be null");
            }
            IMessageFactory mf = (IMessageFactory) _mimeToFactoryMap[mimeType];
            if (mf == null)
            {
                throw new AMQException("Unsupport MIME type of " + mimeType);
            }
            else
            {
                return mf.CreateMessage();
            }
        }

        /// <summary>
        /// Construct a new registry with the default message factories registered
        /// </summary>
        /// <returns>a message factory registry</returns>
        public static MessageFactoryRegistry NewDefaultRegistry()
        {
            MessageFactoryRegistry mf = new MessageFactoryRegistry();
            mf.RegisterFactory("text/plain", new XMSTextMessageFactory());
            mf.RegisterFactory("text/xml", new XMSTextMessageFactory());
            mf.RegisterFactory("application/octet-stream", new XMSBytesMessageFactory());
            // TODO: use bytes message for default message factory
            //if (FieldTable.grm)
            mf.RegisterFactory(null, new XMSBytesMessageFactory());
            return mf;
        }
    }
}
